### PR TITLE
Make HdrHistogram_c atomic so it can be used in a multi threaded context

### PR DESCRIPTION
**Pull request by rdemellow**

### CODE REVIEW

File: src/hdr_histogram.c [hunk 1]
```diff
@@ -367,10 +386,12 @@ int hdr_alloc(int64_t highest_trackable_value, int significant_figures, struct h
 /* reset a histogram to zero. */
 void hdr_reset(struct hdr_histogram *h)
 {
+    hdr_mutex_lock(&hdrMutex);
```
Comment (daverigby) [inline src/hdr_histogram.c]: Note: Given that `total_count`, `min_value` and `max_value` are all atomic (and moreover you access them without a mutex elsewhere), you might as well constrain the mutex scope to just around the `memset()` of `h->counts`.

File: src/hdr_histogram.c [hunk 2]
```diff
@@ -50,7 +55,7 @@ static int32_t normalize_index(const struct hdr_histogram* h, int32_t index)
 
 static int64_t counts_get_direct(const struct hdr_histogram* h, int32_t index)
 {
-    return h->counts[index];
+    return atomic_load_explicit((atomic_int_least64_t*)&h->counts[index], memory_order_relaxed);
```
Comment (daverigby) [inline src/hdr_histogram.c]: I don't believe you need the `(atomic_int_least64_t*)` cast here, given that already the type of `h->counts`.

File: src/hdr_histogram.c [hunk 3]
```diff
@@ -62,14 +67,20 @@ static void counts_inc_normalised(
...
+    atomic_fetch_add_explicit(&h->total_count, value, memory_order_relaxed);
 }
 
 static void update_min_max(struct hdr_histogram* h, int64_t value)
 {
-    h->min_value = (value < h->min_value && value != 0) ? value : h->min_value;
-    h->max_value = (value > h->max_value) ? value : h->max_value;
+    int64_t min = atomic_load_explicit(&h->min_value, memory_order_acquire);
+    if (value < min && value != 0) {
+        atomic_store_explicit(&h->min_value, value, memory_order_release);
```
Comment (daverigby) [inline src/hdr_histogram.c]: Note this isn't thread-safe - another thread might have already updated min_value between the load at line 76 and the store at line 78.

You should instead use a compare_exchange loop - see https://en.cppreference.com/w/c/atomic/atomic_compare_exchange

File: src/hdr_histogram.c [hunk 4]
```diff
@@ -62,14 +67,20 @@ static void counts_inc_normalised(
...
+    }
+    int64_t max = atomic_load_explicit(&h->max_value, memory_order_acquire);
+    if (value > max) {
+        atomic_store_explicit(&h->max_value, value, memory_order_release);
```
Comment (daverigby) [inline src/hdr_histogram.c]: Same here.

File: src/hdr_histogram.c [hunk 5]
```diff
@@ -337,17 +353,20 @@ int hdr_init(
         return r;
     }
 
+    hdr_mutex_lock(&hdrMutex);
```
Comment (daverigby) [inline src/hdr_histogram.c]: Note that until this function returns the caller cannot get access to the newly-created histogram object - and hence there's no possibility another thread could be manipulating it. 

As such I don't believe you need to acquire the mutex in this function.

File: src/hdr_histogram.c (hunk 1 above)
Comment (daverigby) [reply src/hdr_histogram.c]: Furthermore, given h->counts can again be modified without a lock by other threads , you could just remove the mutex entirely and simply clear counts using a for() loop using atomic_store_explicit() on each element.

**Review by daverigby — CHANGES_REQUESTED** [review]

File: src/hdr_histogram.c (hunk 2 above)
Comment (rdemellow) [reply src/hdr_histogram.c]: Done

File: src/hdr_histogram.c (hunk 5 above)
Comment (rdemellow) [reply src/hdr_histogram.c]: Thanks Dave good point (bit of a derp here)

File: src/hdr_histogram.c (hunk 3 above)
Comment (rdemellow) [reply src/hdr_histogram.c]: Right miss understood the relationship between memory_order_aquire and memory_order_release will fix

File: src/hdr_histogram.c (hunk 1 above)
Comment (rdemellow) [reply src/hdr_histogram.c]: Done

File: src/hdr_histogram.c (hunk 5 above)
Comment (rdemellow) [reply src/hdr_histogram.c]: Done

File: src/hdr_histogram.c (hunk 3 above)
Comment (rdemellow) [reply src/hdr_histogram.c]: used a mutex as there's no atomic less than or greater than comparison and set

File: src/hdr_histogram.c [hunk 6]
```diff
@@ -13,10 +14,14 @@
 #include <stdint.h>
 #include <errno.h>
 #include <inttypes.h>
+#include <hdr_thread.h>
 
 #include "hdr_histogram.h"
 #include "hdr_tests.h"
 
+static hdr_mutex hdrMinMaxMutex = {0};
```
Comment (rdemellow) [inline src/hdr_histogram.c]: this mutex should be destroyed in the hdr_close

File: src/hdr_histogram.c [hunk 7]
```diff
@@ -62,14 +63,34 @@ static void counts_inc_normalised(
...
+                atomic_load_explicit(&h->min_value, memory_order_relaxed);
+
+        do {
+            updateValue = false;
+            if (value != 0 && value < observedMin) {
+                updateValue = true;
+            }
+        } while (updateValue && !atomic_compare_exchange_weak(
+                                        &h->min_value, &observedMin, value));
+    }
```
Comment (jimwwalker) [inline src/hdr_histogram.c:81]: Can be reduced to something like:

```
int64_t observedMin =
                atomic_load_explicit(&h->min_value, memory_order_relaxed);
while (value > 0 && value < observedMin && !atomic_compare_exchange_weak(&h->min_value, &observedMin, value)) {
}
```

Or 
```
if (value > 0) {
        int64_t observedMin = atomic_load_explicit(&h->min_value, memory_order_relaxed);
        while (value < observedMin && !atomic_compare_exchange_weak(&h->min_value, &observedMin, value)) {

        }     
    }
```

Similar for max

**Review by jimwwalker — CHANGES_REQUESTED** [review]

File: src/hdr_histogram.c (hunk 7 above)
Comment (daverigby) [reply src/hdr_histogram.c:81]: Nit: You probably want the _explicit variant of compare_exchange_weak so you can specify relaxed memory ordering.

Comment (rdemellow) [reply src/hdr_histogram.c:81]: Done

**Review by jimwwalker — APPROVED** [review]

**Review by daverigby — APPROVED** [review]

**Pull request merged**